File: solution-py/solution-py/src/solver/sat/dimacs_parser.py
import os

from sat_instance import SATInstance

def parse_cnf_file(file_name):
    sat_instance = None
    try:
        with open(file_name, 'r') as f:
            lines = f.readlines()

        # skip comment lines
        idx = 0
        while idx < len(lines) and lines[idx].strip().startswith('c'):
            idx += 1

        tokens = lines[idx].split()
        if tokens[0] != 'p':
            raise ValueError("Error: DIMACS file does not have problem line")
        if tokens[1] != 'cnf':
            print("Error: DIMACS file format is not cnf")
            return sat_instance

        num_vars = int(tokens[2])
        num_clauses = int(tokens[3])
        sat_instance = SATInstance(num_vars, num_clauses)
        idx += 1

        while idx < len(lines):
            line = lines[idx].strip()
            idx += 1
            if line.startswith('c') or not line:
                continue
            clause_tokens = line.split()
            if clause_tokens[-1] != '0':
                raise ValueError(f"Error: clause line does not end with 0 {clause_tokens}")
            clause = set()
            for lit_str in clause_tokens[:-1]:
                lit = int(lit_str)
                clause.add(lit)
                sat_instance.add_variable(lit)
            sat_instance.add_clause(clause)

    except FileNotFoundError:
        raise FileNotFoundError(f"Error: DIMACS file not found {file_name}")

    return sat_instance

File: solution/solution-b/src/dimacs_parser.cpp
#include "dimacs_parser.h"

#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <unordered_set>
#include <vector>

using namespace std;

SATInstance parseCNFFile(const string &fileName) {
    ifstream file(fileName);
    if (!file.is_open()) {
        throw runtime_error("Error: DIMACS file not found: " + fileName);
    }

    string line;
    vector<string> tokens;

    // Skip comment lines (lines starting with 'c')
    while (getline(file, line)) {
        istringstream iss(line);
        tokens.clear();
        string token;
        while (iss >> token)
            tokens.push_back(token);

        if (!tokens.empty() && tokens[0] != "c") {
            break;
        }
    }

    // Ensure first non-comment line is the problem line
    if (tokens.empty() || tokens[0] != "p") {
        throw invalid_argument("Error: DIMACS file does not have a valid problem line.");
    }

    if (tokens.size() < 4 || tokens[1] != "cnf") {
        throw runtime_error("Error: DIMACS file format is not CNF.");
    }

    int numVars = stoi(tokens[2]);
    int numClauses = stoi(tokens[3]);
    SATInstance satInstance(numVars, numClauses);

    // Parse clauses
    vector<int> clause;
    while (getline(file, line)) {
        istringstream iss(line);
        tokens.clear();
        string token;
        while (iss >> token)
            tokens.push_back(token);

        if (tokens.empty() || tokens[0] == "c") {
            continue; // Skip comments
        }

        if (tokens.back() != "0") {
            throw invalid_argument("Error: Clause line does not end with 0.");
        }

        // Process literals in clause
        for (size_t i = 0; i < tokens.size() - 1; ++i) {
            if (tokens[i].empty())
                continue;

            int literal = stoi(tokens[i]);
            clause.push_back(literal);
        }

        // Add clause to SATInstance
        satInstance.addClause(clause);
        clause.clear();
    }

    return satInstance;
}
